File: library-management/app/schemas/borrowing.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.borrowing import BorrowStatus


class BorrowBookRequest(BaseModel):
    member_id: int = Field(..., gt=0)
    book_id: int = Field(..., gt=0)


class ReturnBookRequest(BaseModel):
    borrowing_id: int = Field(..., gt=0)


class BorrowResponse(BaseModel):
    id: int
    member_id: int
    book_id: int
    borrow_date: datetime
    due_date: datetime
    return_date: Optional[datetime]
    status: BorrowStatus

    model_config = ConfigDict(from_attributes=True)

File: frontend/types/borrowing.ts
// frontend/types/borrowing.ts

/**
 * Borrowing status returned by the backend.
 */
export type BorrowStatus =
  | "BORROWED"
  | "RETURNED"
  | "OVERDUE";

/**
 * Borrowing record.
 */
export interface Borrowing {
  id: number;
  member_id: number;
  book_id: number;

  borrow_date: string;

  due_date: string;

  return_date?: string;

  status: BorrowStatus;
}

/**
 * Borrow request payload.
 */
export interface BorrowBookRequest {
  member_id: number;
  book_id: number;
}

/**
 * Return book request payload.
 */
export interface ReturnBookRequest {
  borrowing_id: number;
}

/**
 * Response from borrow/return API.
 */
export type BorrowResponse = Borrowing;